eventsources/sources/redisStream/start.go

```diff
@@ -0,0 +1,215 @@
+		}
+		opt.Password = password
+	}
+
+	if redisEventSource.TLS != nil {
+		tlsConfig, err := common.GetTLSConfig(redisEventSource.TLS)
+		if err != nil {
+			return errors.Wrap(err, "failed to get the tls configuration")
+		}
+		opt.TLSConfig = tlsConfig
+	}
+
+	log.Infof("setting up a redis client for %s...", redisEventSource.HostAddress)
+	client := redis.NewClient(opt)
+
+	if status := client.Ping(ctx); status.Err() != nil {
+		return errors.Wrapf(status.Err(), "failed to connect to host %s and db %d for event source %s", redisEventSource.HostAddress, redisEventSource.DB, el.GetEventName())
+	}
+	log.Infof("connected to redis server %s", redisEventSource.HostAddress)
+
+	// Create a common consumer group on all streams to start reading from beginning of the streams.
+	// Only proceeds if all the streams are already present
+	consumersGroup := "argo-events-cg"
+	if len(redisEventSource.ConsumerGroup) != 0 {
+		consumersGroup = redisEventSource.ConsumerGroup
+	}
+	for _, stream := range redisEventSource.Streams {
+		// create a consumer group to start reading from the current last entry in the stream (https://redis.io/commands/xgroup-create)
+		if err := client.XGroupCreate(ctx, stream, consumersGroup, "$").Err(); err != nil {
+			// redis package doesn't seem to expose concrete error types
+			if err.Error() != "BUSYGROUP Consumer Group name already exists" {
+				return errors.Wrapf(err, "creating consumer group %s for stream %s on host %s for event source %s", consumersGroup, stream, redisEventSource.HostAddress, el.GetEventName())
+			}
+			log.Infof("Consumer group %q already exists in stream %q", consumersGroup, stream)
+		}
+	}
+
+	readGroupArgs := make([]string, 2*len(redisEventSource.Streams))
+	copy(readGroupArgs, redisEventSource.Streams)
+	// Start by reading our pending messages(previously read but not acknowledged).
+	streamToLastEntryMapping := make(map[string]string, len(redisEventSource.Streams))
+	for _, s := range redisEventSource.Streams {
+		streamToLastEntryMapping[s] = "0-0"
+	}
+
+	updateReadGroupArgs := func() {
+		for i, s := range redisEventSource.Streams {
+			readGroupArgs[i+len(redisEventSource.Streams)] = streamToLastEntryMapping[s]
+		}
+	}
+	updateReadGroupArgs()
+
+	msgCount := redisEventSource.MaxMsgCountPerRead
+	if msgCount == 0 {
+		msgCount = 10
+	}
+
+	var msgsToAcknowledge []string
+	for {
+		select {
+		case <-ctx.Done():
+			log.Infof("Redis stream event source for host %s is stopped", redisEventSource.HostAddress)
+			return nil
+		default:
+		}
+		entries, err := client.XReadGroup(ctx, &redis.XReadGroupArgs{
+			Group:    consumersGroup,
+			Consumer: "argo-events-worker",
+			Streams:  readGroupArgs,
+			Count:    int64(msgCount),
+			Block:    2 * time.Second,
+			NoAck:    false,
+		}).Result()
+		if err != nil {
+			if err == redis.Nil {
+				continue
+			}
+			return errors.Wrapf(err, "reading streams %s using XREADGROUP", strings.Join(redisEventSource.Streams, ", "))
```

Comment:
Should we just log the error and continue? For example, if there's a network connection issue, and it fails to read, we should let it go and retry right?

Reply:
That makes sense. Made the change.